server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java

```diff
@@ -997,6 +1011,51 @@ public void testClusterCannotFormWithFailingJoinValidation() {
         assertTrue(cluster.clusterNodes.stream().allMatch(cn -> cn.getLastAppliedClusterState().version() == 0));
     }
 
+    public void testCannotJoinClusterWithDifferentUUID() throws IllegalAccessException {
+        final Cluster cluster1 = new Cluster(randomIntBetween(1, 3));
+        cluster1.runRandomly();
+        cluster1.stabilise();
+
+        final Cluster cluster2 = new Cluster(3);
+        cluster2.runRandomly();
+        cluster2.stabilise();
+
+        final ClusterNode shiftedNode = randomFrom(cluster2.clusterNodes).restartedNode();
+        final ClusterNode newNode = cluster1.new ClusterNode(nextNodeIndex.getAndIncrement(),
+            shiftedNode.getLocalNode(), n -> shiftedNode.persistedState);
+        cluster1.clusterNodes.add(newNode);
+
+        MockLogAppender mockAppender = new MockLogAppender();
```

Comment:
+1 to that. Why can't we e.g. call `cluster1.runFor(DEFAULT_STABILISATION_TIME)` and then assert that the node didn't manage to join the cluster?

Reply:
I've extended it to `DEFAULT_STABILISATION_TIME`, but I also want to make sure it fails for the right reason, hence keeping the logging check.

Reply:
I'm all ears if you have better suggestions.

Reply:
How about asserting that if we change nothing except the cluster UUID on disk then the node _does_ join the cluster?

Reply:
But there is no guarantee that it will if it has same term but higher version than master? I think these kinds of tests will be more useful after having the detach-cluster tool.